### ISSUE TITLE
dtype of mixed ints and strings depends on the element order

### ISSUE THREAD

**Issue by btel** [numpy/numpy#7973]
The dtype of NumPy array created from a Python list containing both ints and strings depends on the order of the elements:

``` python
In [3]: np.array(['a', 1]).dtype
Out[3]: dtype('<U1')

In [4]: np.array([1, 'a']).dtype
Out[4]: dtype('<U21')
```

tested on Python 3.5.2 and NumPy 1.11.1

Note that it's related (but not the same) to issue #6550

**Comment by btel** [comment]
Note that the type promotion is consistent with the second case:

```
In [13]: np.promote_types('i8', '<U1')
Out[13]: dtype('<U21')
````

I think the result of the first example (`['a', 1]`) is due to the optimization in [PyArray_DTypeFromObjectHelper](https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/common.c#L177). It seems that type promotion is not used when the int get be encoded as a string with the size determined from the previous elements: 


```c
            if (*out_dtype != NULL &&
                    (*out_dtype)->type_num == string_type &&
                    (*out_dtype)->elsize >= itemsize) {
                return 0;
```

https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/common.c#L248

**Comment by kurtqq** [comment]
@btel was this eventually fixed?

**Comment by btel** [comment]
apparently not yet

```
>>> import numpy as np
>>> np.array(['a', 1]).dtype
dtype('<U1')
>>> np.array([1, 'a']).dtype
dtype('<U21')
>>> np.__version__
'1.19.2'
```

**Comment by seberg** [comment]
Its fixed *and* deprecated in 1.20, you have to use `np.array([1, "a"], dtype="U")`, which will give you the identical result in both cases to begin with. The result is than always length 1, just because that is the more common variant currently.

**Comment by seberg** [comment]
Sorry, its a bit "weirder"... Just to be clear: Right now, its deprecated, but you get `U21` *always*.  If you use `dtype="U"` (which you should and will have to), the result is unchanged and gives you `U1` always.
